primitives/arithmetic/src/lib.rs

```diff
@@ -85,8 +86,328 @@ where
+/// element with `diff / input.len()` might easily saturate and results will be `Err`.
+pub fn normalize<T>(input: &[T], targeted_sum: T) -> Result<Vec<T>, &'static str>
+	where T: Clone + Copy + Ord + BaseArithmetic + Debug,
+{
+	let mut sum = T::zero();
+	for t in input.iter() {
+		sum = sum.checked_add(t).ok_or("sum of input cannot fit in `T`")?;
+	}
+	let count = input.len();
+
+	// early exit if needed.
+	if count.is_zero() {
+		return Ok(Vec::<T>::new());
+	}
+
+	let diff = targeted_sum.max(sum) - targeted_sum.min(sum);
+	if diff.is_zero() {
+		return Ok(input.to_vec());
+	}
+
+	let needs_bump = targeted_sum > sum;
+	let per_round = diff / count.try_into().map_err(|_| "failed to build T from usize")?;
+	let mut leftover = diff % count.try_into().map_err(|_| "failed to build T from usize")?;
+
+	let mut output_with_idx = input.iter().cloned().enumerate().collect::<Vec<(usize, T)>>();
+	// sort output once based on diff. This will require more data transfer and saving original
+	// index, but we sort only twice instead: once now and once at the very end.
+	output_with_idx.sort_unstable_by_key(|x| x.1);
+
+	if needs_bump {
+		// must increase the values a bit. Bump from the min element. Index of minimum is now zero
+		// because we did a sort. If at any point the min goes greater or equal the `max_threshold`,
+		// we move to the next minimum.
+		let mut min_index = 0;
+		// at this threshold we move to next index.
+		let threshold = output_with_idx
+			.last()
+			.expect("length of input is greater than zero; it must have a last; qed").1;
+
+		if !per_round.is_zero() {
+			for _ in 0..count {
+				output_with_idx[min_index].1 = output_with_idx[min_index].1
+					.checked_add(&per_round)
+					.ok_or("Failed to add.")?;
+				if output_with_idx[min_index].1 >= threshold {
+					min_index += 1;
+					min_index = min_index % count;
+				}
+			}
+		}
+
+		// continue with the previous min_index
+		while !leftover.is_zero() {
+			output_with_idx[min_index].1 = output_with_idx[min_index].1
+				.checked_add(&One::one())
+				.ok_or("Failed to add.")?;
+			if output_with_idx[min_index].1 >= threshold {
+				min_index += 1;
+				min_index = min_index % count;
+			}
+			leftover -= One::one()
+		}
+	} else {
+		// must decrease the stakes a bit. decrement from the max element. index of maximum is now
+		// last. if at any point the max goes less or equal the `min_threshold`, we move to the next
+		// maximum.
+		let mut max_index = count - 1;
+		// at this threshold we move to next index.
+		let threshold = output_with_idx
+			.first()
+			.expect("length of input is greater than zero; it must have a fist; qed").1;
+
+		if !per_round.is_zero() {
+			for _ in 0..count {
+				output_with_idx[max_index].1 = output_with_idx[max_index].1
+					.checked_sub(&per_round)
+					.ok_or("Failed to subtract.")?;
+				if output_with_idx[max_index].1 <= threshold {
+					max_index = max_index.checked_sub(1).unwrap_or(count - 1);
+				}
+			}
```

Comment:
this part looks wrong to me: if input is `[1, 8, 9]` and target is 1 then
* diff is 17, per_round is `17/3=5`
* first loop give `[1, 8, 4]`
* second loop give "Failed to subtract`.

I think this behavior will be very unexpected from caller and the documentation only says:

>  Moreover, if any of the internal operations saturate

Which is not enough I think.

Better would be to handle it correctly without failing for such inputs.

Reply:
So the reason that I did it as such was that I was worried that handling it correctly would be too complicated. But let me try. How about this: 

1. Given `per_round = p` and `remainder = r`
2. we loop over elements from the min-index, and to a checked_sub
  2.1 if `Some(_)`, then we move on. 
  2.2 if `None`, we will move to the next index. 

This will make the implementation slightly more complicated but will cover your example.

Reply:
I am giving this a try, but as expected, it is much more complicated for the subtraction side.

Reply:
I feel like in order to work in all case you have to handle the case where you split per_round, in my example you can go to `[1, 1, 1]` by just subtracting 5 to elements

Reply:
We settled for a simpler case that we add add the error values to leftover. Slow, but accurate. And it should barely ever happen.